akka-http-core/src/main/scala/akka/http/impl/engine/http2/framing/FrameRenderer.scala

```diff
@@ -24,141 +21,200 @@ private[http2] object FrameRenderer {
+        HeaderBuilder(data.size)
+          .putHeader(
+            Http2Protocol.FrameType.PING,
+            Http2Protocol.Flags.ACK.ifSet(ack),
+            Http2Protocol.NoStreamId
+          )
+          .put(data)
+          .result()
 
       case RstStreamFrame(streamId, errorCode) =>
-        renderFrame(
-          Http2Protocol.FrameType.RST_STREAM,
-          Http2Protocol.Flags.NO_FLAGS,
-          streamId,
-          new ByteStringBuilder().putInt(errorCode.id).result()
-        )
+        HeaderBuilder(4)
+          .putHeader(
+            Http2Protocol.FrameType.RST_STREAM,
+            Http2Protocol.Flags.NO_FLAGS,
+            streamId
+          )
+          .putInt32(errorCode.id)
+          .result()
 
       case PushPromiseFrame(streamId, endHeaders, promisedStreamId, headerBlockFragment) =>
-        renderFrame(
-          Http2Protocol.FrameType.PUSH_PROMISE,
-          Http2Protocol.Flags.END_HEADERS.ifSet(endHeaders),
-          streamId,
-          new ByteStringBuilder()
-            .putInt(promisedStreamId)
-            .append(headerBlockFragment)
-            .result()
-        )
-
-      case frame @ PriorityFrame(streamId, exclusiveFlag, streamDependency, weight) =>
-        renderFrame(
-          Http2Protocol.FrameType.PRIORITY,
-          Http2Protocol.Flags.NO_FLAGS,
-          streamId,
-          renderPriorityInfo(frame)
-        )
+        HeaderBuilder(4 + headerBlockFragment.size)
+          .putHeader(
+            Http2Protocol.FrameType.PUSH_PROMISE,
+            Http2Protocol.Flags.END_HEADERS.ifSet(endHeaders),
+            streamId
+          )
+          .putInt32(promisedStreamId)
+          .put(headerBlockFragment)
+          .result()
+
+      case frame @ PriorityFrame(streamId, _, _, _) =>
+        HeaderBuilder(5)
+          .putHeader(
+            Http2Protocol.FrameType.PRIORITY,
+            Http2Protocol.Flags.NO_FLAGS,
+            streamId
+          )
+          .putPriorityInfo(frame)
+          .result()
       case _ => throw new IllegalStateException(s"Unexpected frame type ${frame.frameTypeName}.")
     }
 
-  def renderFrame(tpe: FrameType, flags: ByteFlag, streamId: Int, payload: ByteString): ByteString = {
-    val length = payload.length
-    val headerBytes = new Array[Byte](9)
-    headerBytes(0) = (length >> 16).toByte
-    headerBytes(1) = (length >> 8).toByte
-    headerBytes(2) = (length >> 0).toByte
-    headerBytes(3) = tpe.id.toByte
-    headerBytes(4) = flags.value.toByte
-    headerBytes(5) = (streamId >> 24).toByte
-    headerBytes(6) = (streamId >> 16).toByte
-    headerBytes(7) = (streamId >> 8).toByte
-    headerBytes(8) = (streamId >> 0).toByte
-
-    ByteString1C(headerBytes) ++ payload
+  private class HeaderBuilder(payloadSize: Int) {
+    private val targetSize = 9 + payloadSize
+    private val buffer = new Array[Byte](targetSize)
+    private var pos = 0
+
+    def putHeader(tpe: FrameType, flags: ByteFlag, streamId: Int): HeaderBuilder = {
```

Comment:
Since there is always a header, maybe pass these in in the constructor rather than via a `putHeader`?

Reply:
:+1: done